### PR TITLE
Refactor sorting

### PR DESCRIPTION
**Pull request by Arif-Khalid**
### Summary:

Fixes part of #249 

#### Type of change:

- 🎨 Code Refactoring

### Changes Made:

- Refactored sorting behaviour of IssueDataTable into dropdownFilters
  - Removed reference of MatSort in IssueDataTable
  - Replaced with matsort Sort object stored in dropdown filters
  - Removed unused code

### Proposed Commit Message:

```
MatSort is passed by reference into IssueDataTables,
lifted up from filter-bar component. Changes to 
MatSort are detected and handled in IssueDataTables.

MatSort reference in IssueDataTables performs the 
responsibility of a dropdown filter outside of the 
filters service.
This violates good programming practices.

Let's refactor the sort filter into the filters service.
```

<details><summary>
<h3>Checklist:</h3>
</summary>

- [x] I have tested my changes thoroughly.
- [x] I have created tests for any new code files created in this PR or provided a link to a issue/PR that addresses this.
- [x] I have added or modified code comments to improve code readability where necessary.
- [x] I have updated the project's documentation as necessary.

</details>

### CODE REVIEW

**Review by nknguyenhc — CHANGES_REQUESTED** [review]
image

Filter on labels do not work. Any other filters similarly do not work.

File: src/app/shared/issue-tables/dropdownfilter.ts [hunk 1]
```diff
@@ -21,10 +23,10 @@ export const DEFAULT_DROPDOWN_FILTER = <DropdownFilter>{
...
- * This functions returns a function to check if a issue matches a dropdownfilter
+ * This functions returns the data passed in after all the filters of dropdownFilters are applied
  */
-export function applyDropdownFilter(dropdownFilter: DropdownFilter): (a: Issue) => boolean {
-  return (issue) => {
+export function applyDropdownFilter(dropdownFilter: DropdownFilter, data: Issue[]): Issue[] {
+  const filteredData: Issue[] = data.filter((issue) => {
     let ret = true;
 
     if (dropdownFilter.status === 'open') {
```
[the remark this answers is not in the document]
Comment (Arif-Khalid) [reply src/app/shared/issue-tables/dropdownfilter.ts:32]: Excellent catch. I must have had uncommited local changes. The error has been fixed and all filters should be working now.

**Review by nknguyenhc — DISMISSED** [review]
LGTM!

File: src/app/shared/issue-tables/dropdownfilter.ts [hunk 2]
```diff
@@ -45,5 +47,6 @@ export function applyDropdownFilter(dropdownFilter: DropdownFilter): (a: Issue)
     ret = ret && dropdownFilter.milestones.some((milestone) => issue.milestone.number === milestone);
 
     return ret && dropdownFilter.labels.every((label) => issue.labels.includes(label));
-  };
+  });
+  return getSortedData(dropdownFilter.sort, filteredData);
```
Comment (vigneshsankariyer1234567890) [inline src/app/shared/issue-tables/dropdownfilter.ts:51]: Makes sense to apply filter then sort the data, but `applyDropdownFilter` only has one consumer in `IssuesDataTable::loadIssues` and `getSortedData` only has one consumer in `applyDropdownFilter`. Can we ensure single responsibility by just having `applyDropdownFilter` apply the filters, and after the filters have been applied, apply the sort directly in `IssuesDataTable::loadIssues`?

File: src/app/shared/issue-tables/issue-sorter.ts [hunk 3]
```diff
@@ -1,8 +1,8 @@
-import { MatSort } from '@angular/material/sort';
+import { Sort } from '@angular/material/sort';
 import * as moment from 'moment';
 import { Issue } from '../../core/models/issue.model';
 
-export function getSortedData(sort: MatSort, data: Issue[]): Issue[] {
+export function getSortedData(sort: Sort, data: Issue[]): Issue[] {
```
Comment (vigneshsankariyer1234567890) [inline src/app/shared/issue-tables/issue-sorter.ts:5]: Perhaps we can rename this to `applySort`, in the same convention as applying filters in `./dropdownfilter.ts`?

File: src/app/shared/issue-tables/IssuesDataTable.ts [hunk 4]
```diff
@@ -87,12 +78,10 @@ export class IssuesDataTable extends DataSource<Issue> implements FilterableSour
...
+
           // Dropdown Filters
-          data = data.filter(applyDropdownFilter(this.dropdownFilter));
+          data = applyDropdownFilter(this.dropdownFilter, data);
 
-          if (this.sort !== undefined) {
-            data = getSortedData(this.sort, data);
-          }
           data = applySearchFilter(this.filter, this.displayedColumn, this.issueService, data);
           this.count = data.length;
```
Comment (vigneshsankariyer1234567890) [inline src/app/shared/issue-tables/IssuesDataTable.ts:87]: Over here, could we do `applyDropdownFilter`, then `applySearchFilter` and finally `applySort`?

**Review by vigneshsankariyer1234567890 — CHANGES_REQUESTED** [review]
Overall, slick PR and nice changes to integrate changes from filters! Just a few comments regarding SRP! 😄

File: src/app/shared/issue-tables/dropdownfilter.ts (hunk 2 above)
Comment (Arif-Khalid) [reply src/app/shared/issue-tables/dropdownfilter.ts]: I see what you mean. I have applied the recommended changes to perform sorting in ``IssuesDataTable::loadIssues``.

File: src/app/shared/issue-tables/IssuesDataTable.ts (hunk 4 above)
Comment (Arif-Khalid) [reply src/app/shared/issue-tables/IssuesDataTable.ts:88]: Order of applying different filters and sorting has been updated.

**Review by vigneshsankariyer1234567890 — APPROVED** [review]
Nicely done, LGTM! 🚀

**Pull request merged**